Workflow Automation with Github Actions

alyip98 opened the pull request:
[Preview](https://deploy-preview-187--learningresources.netlify.com/contents/projectmanagement/githubactions)

As this is a pretty rough draft, I would appreciate reviews that focus on the quality of the content and its usefulness to potential readers. 

Todo:
- [ ] ~Refactor to minimize screenshots~
- [x] Upload images to repo
- [x] Recommend useful actions such as cache
- [x] Add Table of Contents
- [x] Title Case
- [x] Github --> GitHub
- [x] How to Get Started with GitHub Actions
- [x] Point out shortcomings with the official doc in a nice way
- [ ] ~Condense benefits into list OR expand on them?~
- [ ] ~Add intro to ## How is GitHub Actions Being Used~ Section removed
- [x] Reduce repetition
- [x] Change list indexes to 1.
- [x] Remove numbers reference for runner mins

alyip98 committed:
Add github actions article

damithc commented:
Note that GitHub Actions is very new and likely to evolve rapidly in the coming months. Try to write in a way the contents stays correct for some time e.g., minimize screenshots.

damithc commented:
> I can't request reviews

Probably didn't accept the invitation to join the org last time. I've resent the invitation.

alyip98 committed:
Apply template

le0tan commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+## Why Github Actions?
+<!--Motivate the reader to learn about X. Describe benefits of X to make the reader interested in X. Try to give a balanced view of by mentioning also WHY NOT X i.e., mention both advantages and disadvantages.-->
+### Easy to use
+Setting up a workflow in a repository is easy and hassle-free. Templates for popular languages are available and the built-in editor offers code completion and documentation at a glance.
+![](https://i.imgur.com/JiVXlO4.png)
+
+
+
+### Reusable actions
+Actions are a small, reusable unit in a workflow. When setting up a workflow, we can use actions written by the community. The marketplace allows users to browse actions written by others for use in our workflows.
+
+
+![](https://i.imgur.com/m9PddIu.png)
+
+
+
+### Cross OS support
+Github Actions gives you the option to [run workflows](https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources) on Ubuntu, Windows and macOS.
+
+### Docker support
+Github Actions provides the ability to run docker images for projects that use a dockerized environment.
+
+### Powerful
+Integration with the Github ecosystem allows Github Actions to do more than just build/test/deploy. You can automate everything related to your Github project, including bug triaging, labeling, automated dependency updating etc.
+
+### Affordable
+The free plan offers 2000 minutes of build time per month for public repositories and a pay-per-minute plan for further use.
+
+---
+
+## How does Github Actions work
```
Since these benefits' descriptions are relatviely short, probably consider using numbered list instead of titles?
e.g. `1. **Affordable**: ...`

le0tan commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+### Set up a simple workflow (Node.js)
+Fortunately for us, Github has made it extremely easy to set up a simple workflow without even leaving their website. 
+
+![](https://i.imgur.com/zC4ScYn.png)
```
Consider pushing the static asset to the repo instead of linking to an external service.

le0tan commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+1. Head to the Actions tab on your repository
+2. You can choose to set up a workflow from a template or from scratch. We will use the Node.js template in this example
+3. The online editor is very powerful and easy to use, boasting features such as inline documentation and autocomplete
+![](https://i.imgur.com/gknkruJ.png)
```
same as the last one, consider using static asset within repo

le0tan commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+    6. Report the status of the workflow via a status badge
+    ![](https://i.imgur.com/3Tsm2Ii.png)
+5. Since we set the workflow up on the `master` branch, when we commit the newly-created workflow file, we can see Github Actions in action by navigating to the Actions tab
+    ![](https://i.imgur.com/ZxBlv8d.png)
```
same here

le0tan commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+
+
+## Where to Go from Here?
```
Probably you may recommend some widely used actions like [cache](https://github.com/actions/cache), etc.

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+<div class="website-content">
+
+{{ booktitle | safe }}
+
```
can add a table of contents here, refer to other PRs for examples

alyip98 replied:
Just curious, is there a real need for this aside from being consistent, since the right navbar doubles as a ToC?

jamessspanggg replied:
oh it will be responsive based on screen size, it will not show until ur screen is of a certain smaller size.

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+🔎 
+
+
+## What is Github Actions?
```
Change all Github -> GitHub

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+
+### Reusable actions
+Actions are a small, reusable unit in a workflow. When setting up a workflow, we can use actions written by the community. The marketplace allows users to browse actions written by others for use in our workflows.
```
```suggestion
Actions are small, reusable units in a workflow. When setting up a workflow, we can re-use actions that were already written by the community. These actions are available at the marketplace.
```

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+
+
+## How to Get Started with X?
```
How to Get Started with GitHub Actions?

Tejas2805 replied:
Whichever way you decide to go ahead, please make sure to do it in titleCase

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+
+
+### Reusable actions
```
Can consider using title case for all headers

Tejas2805 replied:
```suggestion
### Reusable Actions
```

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+## What is Github Actions?
+<!--Explain how X fits into the big picture of SE. Describe it relative to topics the reader is likely to know.-->
+Github Actions is a workflow automation solution that is tightly integrated into the Github ecosystem. It is commonly used to provide end-to-end continuous integration and continuous deployment (CI/CD) directly in your repository. With its tight integration with Github's ecosystem of services, users have the potential to automate more than just CI/CD.
+
```
Perhaps you can define the `workflow`, `actions` and `triggers` here instead of at How Does GitHub Actions work? Because I think that these are the basic knowledge that the reader should know before diving into the benefits and so on.

Then the How Does GitHub Actions work section should just be Example GitHub Actions Workflow. 

My 2 cents haha

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+<!--
+Provide a learning path for the reader. Try to give one good learning path rather than many random resources.
+-->
+The [official documentation](https://help.github.com/en/actions/getting-started-with-github-actions/overview) provides a useful overview of Github Actions (summarized in this article), but it does not give you a clear and easy to follow path to getting started. This is a mini-guide for you to get started with Github Actions.
```
Not sure if it's a good idea to critique the official documentation...

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+<!--
+Give more resources. Instead of listing a lot of links, provide a brief summary of what value each resource can provide the reader.
+-->
+The official [documentation](https://help.github.com/en/actions) is a good resource to learn more about Github Actions. It is not an easy read as the information is dense and not organised in a linear manner, but it serves its purpose well as documentation.
```
Same here, not sure if a good idea to critique the official documentation

jamessspanggg reviewed (commented):
Some initial comments:

alyip98 committed:
Update contents/projectManagement/githubActions.md

Co-Authored-By: James Pang <jamespang@u.nus.edu>

alyip98 committed:
Add link to starter-workflows

alyip98 committed:
Correct capitalization of GitHub

alyip98 committed:
Standardise headings to use title case

alyip98 committed:
Improve phrasing

Tejas2805 reviewed (changes requested):
We are aiming for consistency across the book. All headings need to be in titleCase. Have left comments accordingly.

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+## Why Github Actions?
+<!--Motivate the reader to learn about X. Describe benefits of X to make the reader interested in X. Try to give a balanced view of by mentioning also WHY NOT X i.e., mention both advantages and disadvantages.-->
+### Easy to use
```
Easy to Use

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+
+
+### Cross OS support
```
```suggestion
### Cross OS Support
```

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+### Cross OS support
+Github Actions gives you the option to [run workflows](https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources) on Ubuntu, Windows and macOS.
+
+### Docker support
```
```suggestion
### Docker Support
```

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+---
+
+## How does Github Actions work
```
```suggestion
## How Does Github Actions Work?
```

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+Workflows can be triggered by events on the Github repository, ranging from opening pull requests or issues, making commits to scheduled events, or even from an external event by calling Github's REST API endpoint.
+
+
+### How is Github Actions being used
```
```suggestion
### How is GitHub Actions Being Used?
```

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+### How is Github Actions being used
+
+#### Simple CI/CD workflow example for a NodeJS app
```
Can you try to shorten the heading? Too big for a heading I think.

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+-->
+The [official documentation](https://help.github.com/en/actions/getting-started-with-github-actions/overview) provides a useful overview of Github Actions (summarized in this article), but it does not give you a clear and easy to follow path to getting started. This is a mini-guide for you to get started with Github Actions.
+
+### Set up a simple workflow (Node.js)
```
```suggestion
### Set up a Simple Workflow (Node.js)
```

Tejas2805 commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,162 @@
+
+
+
+## Where to Go from Here?
```
```suggestion
## Where to Go From Here?
```

alyip98 committed:
Migrate images to local folder

alyip98 committed:
Add table of contents

alyip98 committed:
Re-organise headings

alyip98 committed:
Credit reviewers

alyip98 commented:
Thanks for the reviews! I have addressed all of the style/formatting issues.

@damithc I decided that I would prefer to keep the screenshots as they help break up the wall of text and effectively demonstrate just how simple it is to set up a workflow on GitHub. I do acknowledge that the feature is relatively new and is subject to changes, but GitHub has been rather stable with regards to their UI, and I believe that changes in the near future will not deviate too much from the screenshots.

@le0tan I chose to keep the heading format of the benefits to emphasize their importance within the article. 

Do let me know if you have any further suggestions/improvements

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+
+📝 [Alfred Yip](https://github.com/alyip98)
+
+🔎 [James Pang](https://github.com/jamessspanggg), [Tan Yuanhong](https://github.com/le0tan), [Tejas Bhuwania](https://github.com/Tejas2805)
```
For consistency, don't use emojis and format authors/reviewers similar to other articles.

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+
+## What is GitHub Actions
+<!--Explain how X fits into the big picture of SE. Describe it relative to topics the reader is likely to know.-->
+GitHub Actions is a workflow automation solution that is tightly integrated into the GitHub ecosystem. It is commonly used to provide end-to-end continuous integration and continuous deployment (CI/CD) directly in your repository. With its tight integration with GitHub's ecosystem of services, users have the potential to automate more than just CI/CD.
```
tight integration mentioned twice in the space of two sentences. Can do with just one.

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+GitHub Actions provides the ability to run docker images for projects that use a dockerized environment.
+
+### 5. Powerful
+Integration with the GitHub ecosystem allows GitHub Actions to do more than just build/test/deploy. You can automate everything related to your GitHub project, including bug triaging, labeling, automated dependency updating etc.
```
>Integration with the GitHub ecosystem allows GitHub Actions to do more than just build/test/deploy.

Something very similar was mentioned earlier in the article. Given this is a short chapter, try to avoid repetition.

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+### 6. Affordable
+Like many of its competitors, GitHub Actions provides **free unlimited runner time for public repositories**. 
+
+For private repositories, the free plan offers 2000 minutes of runner time per month and a pay-per-minute plan for additional minutes.
```
Giving specific numbers increase the risk of the article getting outdated. I'm not sure the recent GitHub change to increase the free tier makes this section outdated already, even before it is published. 
Say something like `free plan offers ample runner minutes ...`

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+
+### Triggers
+Workflows can be triggered by events on the GitHub repository, ranging from opening pull requests or issues, making commits to scheduled events, or even from an external event by calling GitHub's REST API endpoint.
+
```
These sections can be bullet points. No need to add a lot of headings.

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+    3. Add `Needs Author Feedback` label
+3. Otherwise,
+    1. Add `Needs Attention` label
+
```
Seems some explanations are missing from this section e.g., there is no intro

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+1. Trigger when a commit/pull request is made to the master branch
+2. Install dependencies
+3. Build
+4. Run tests with multiple versions of Node on Ubuntu
```
It's better to use `1. ` for all items. Markdown generates the correct number for each item.

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+
+1. Head to the Actions tab on your repository
+2. You can choose to set up a workflow from a template or from scratch. We will use the Node.js template in this example
+3. The online editor is powerful and easy to use, boasting features such as inline documentation and autocomplete
```
This item too contains some info mentioned earlier.

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,155 @@
+React Native | [needs-attention](https://github.com/facebook/react-native/blob/master/.github/workflows/needs-attention.yml), [process-label-actions](https://github.com/facebook/react-native/blob/master/.github/workflows/process-label-actions.yml) | Automatically apply labels to newly created issues and adds comments to issues based on labels
+Node.js | [workflows](https://github.com/nodejs/node/tree/master/.github/workflows) | Linters, tests and build jobs
+Chart.js | [CI](https://github.com/chartjs/Chart.js/blob/master/.github/workflows/ci.yml) | Conditional workflow for windows and ubuntu
+Next.js | [workflows](https://github.com/zeit/next.js/tree/canary/.github/workflows) | Complex yet well structured build/test/deploy workflow. Automatically reports bundle size statistics in pull requests and on releases. Scheduled tests that run every 12 hours.
```
I don't think this section adds much value, or can be replaced with just one sentence.

alyip98 replied:
This entire section? I think that readers may learn more from examples of GitHub Actions used in production compared to sandboxed examples such as the one above.

damithc replied:
It's fine if they are supposed to serve as educational examples. Can keep.

damithc reviewed (changes requested):
Added some comments. Feels like the article is not yet 'fully baked'. Reduce repetition. Remove unnecessary headings. Fill in missing 'connecting' explanations. Make the content more compact where possible.

alyip98 committed:
Update githubActions.md

alyip98 commented:
Thank you for the review @damithc. I apologize for the late reply, I decided to put it on hold until my finals were over. 

I've addressed most of the comments you raised. I intentionally repeated the part about GitHub integration as I believe that is its Unique Selling Point and deserves a bit of extra emphasis.

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,143 @@
+
+I have also compiled a list of popular repositories that use GitHub Actions that you can use as a reference. Do note that this is far from a comprehensive list and is more of a sample of how organizations are leveraging GitHub Actions in their development and production workflow.
+
+<panel header="Repositories that use GitHub Actions">
```
This panel title can be more aligned with the purpose. e.g., `Examples of different ways GitHub Actions can be used`

damithc commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,143 @@
+to run JavaScript within the workflow using GitHub's [Octokit](https://octokit.github.io/rest.js/v17/) API. 
+The Octokit API allows you to create workflows that span across GitHub's ecosystem of services.
+
+I have also compiled a list of popular repositories that use GitHub Actions that you can use as a reference. Do note that this is far from a comprehensive list and is more of a sample of how organizations are leveraging GitHub Actions in their development and production workflow.
```
Avoid using `I` as more authors might join in future.

damithc reviewed (changes requested):
Added a couple of comments. Let me know when you have approval from a couple of peer reviewers.

openorclose commented:
Some links in the TOC for mobile don't work

alyip98 committed:
Fix toc links

alyip98 committed:
Rephrase header title

alyip98 committed:
Remove personal pronoun

alyip98 committed:
Clean up reference to removed example

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,142 @@
+
+<box id="article-toc">
+
+* [What is GitHub Actions‎](#what-is-github-actions)
```
can match this ToC with the siteNav's

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,142 @@
+The official [documentation](https://help.github.com/en/actions) is a good resource to learn more about GitHub Actions.
+The [starter workflows](https://github.com/actions/starter-workflows) repository contains good examples of 
+CI workflows for popular languages as well as automation workflows for GitHub.
+Here are a list of [awesome actions](https://github.com/sdras/awesome-actions) that you can use when writing your workflow.
```
*Here is a list

jamessspanggg commented on contents/projectManagement/githubActions.md:
```diff
@@ -0,0 +1,142 @@
+1. The online editor is powerful and easy to use, boasting features such as inline documentation and autocomplete
+![](images/editor.png)
+1. The template workflow will
+    1. Begin when a commit is pushed or a pull request is opened to `master`
```
think can use a different type of ordered list aka (`i`, `ii`) here

alyip98 replied:
I prefer numbered sublists

jamessspanggg reviewed (commented):
LGTM, just some small nits:

alyip98 committed:
Add daryl as reviewer

alyip98 committed:
Fix grammar

alyip98 committed:
Add H2 to TOC
